### ISSUE TITLE
Function projecting 3D joint coordinates Jˆ3D to the image plane with predicted camera parameters.

### ISSUE THREAD

**Issue by tranducanhbk** [mks0601/NeuralAnnot_RELEASE#2]
Hi mks0610!
I am reproduce your paper but i am stuck at got 2D joints images from  3D joint coordinates Jˆ3D to the image plane with predicted camera parameters.
Could u public this function and λwild,θ parameter for training SMPL model
Many thanks

**Comment by mks0601** [comment]
Please refer to https://github.com/mks0601/Hand4Whole_RELEASE/blob/7c3b7fd6ac3e0824210796a2267fe7f8fdc5e968/main/model.py#L77

**Comment by tranducanhbk** [comment]
How about  λwild,θ to caculate smpl parameter loss. 
Can u tell clearly about it?
Many thanks your kindly response

**Comment by mks0601** [comment]
I set it to 1 when calculating the loss on 256x256 image space.

**Comment by tranducanhbk** [comment]
Thank you,
You are very kind and i have one more question.
This is my regularizer loss function
class RegularizeLoss(nn.Module):
    def __init__(self):
        super(RegularizeLoss, self).__init__()
    def forward(self, pose):
        loss = torch.pow(pose,2)
        loss = torch.sum(loss, dim=1)
        return loss
pose is only  3D body rotations or include  3D body global rotation

**Comment by mks0601** [comment]
I did mean instead of sum in here: loss = torch.sum(loss, dim=1)
Also, 3D global rotation should be exlcluded

**Comment by tranducanhbk** [comment]
I trained with own dataset Format MSCOCO (no face and Hand box) 7 epochs.
The convergence of 2D and regularizer is 1.2 and 0.45. 
I used RotationNet from this link to pridict 3D pseudo-GTs https://github.com/mks0601/Hand4Whole_RELEASE/blob/7c3b7fd6ac3e0824210796a2267fe7f8fdc5e968/common/nets/module.py#L24
And this is result.
2D joint in images obtained by 3D joint coordinates and predicted camera parameters look good but 3D render image is bad.
This is 2D result
![AAA](https://user-images.githubusercontent.com/54969565/170930088-af59c795-e33a-4569-8857-ecd9c744ac89.jpg)
this is 3D result
![render_original_img_body](https://user-images.githubusercontent.com/54969565/170930118-d8cb63aa-2160-4dfb-9b92-bd9b45ecf470.jpg)
Please tell me what i mistake.
Thank you so much

**Comment by mks0601** [comment]
I don't know your exact training settongs, but I guess you should increase l2 reg weight

**Comment by tranducanhbk** [comment]
I follow your Hand4Whole  training setup  https://github.com/mks0601/Hand4Whole_RELEASE.
https://github.com/mks0601/Hand4Whole_RELEASE/blob/7c3b7fd6ac3e0824210796a2267fe7f8fdc5e968/main/config.py#L50
this is forward code to get 3D Gts predict
 if cfg.parts == 'body':
            img_feat = self.forward_backbone(inputs, self.backbone)
            joint_img = smpl.reduce_joint_set(targets['joint_img'])
            smpl_root_pose, smpl_body_pose, smpl_shape, cam_trans = self.forward_rotation_net(img_feat, joint_img, self.rotation_net)
            joint_proj, joint_cam, mesh_cam = self.get_coord({'root_pose': smpl_root_pose, 'body_pose': smpl_body_pose, 'shape': smpl_shape, 'cam_trans': cam_trans}, mode)
            smpl_body_pose = smpl_body_pose.view(-1,(smpl.orig_joint_num-1)*3)
            smpl_pose = torch.cat((smpl_root_pose, smpl_body_pose),1)
and this is loss caculate :
loss['joint_proj'] = self.coord_loss(joint_proj, targets['joint_img'][:,:,:2], meta_info['joint_trunc'])
loss['regularizer'] = self.regular_loss(smpl_body_pose)

this is backward:
loss = {k:loss[k].mean() for k in loss}
sum(loss[k] for k in loss).backward()

Anyway i will increase L2 reg weight (set to 2 )  and test again.

**Comment by tranducanhbk** [comment]
After set L2 reg weight =1 and caculate 2D loss with (x,y) cordinate resize to size (8x8) i got this results
![render_original_img_body](https://user-images.githubusercontent.com/54969565/171227101-d7260e2d-e681-4300-b37a-0d4e24612069.jpg)
 a little bit better but not enogh good. 
Do u have any idead for this issue?

**Comment by mks0601** [comment]
Did you apply the L2 reg on the SMPL shape parameter? I applied L2 reg with weight 1 to the SMPL shape parameter (beta)

**Comment by mks0601** [comment]
Also, I calculated 2D joint loss in 256x256 space.

**Comment by tranducanhbk** [comment]
I did not  apply L2 reg on the SMPL shape parameter, I only caculaye loss with 3D body rotations (θ).
i will try add SMPL shape parameter to caculate loss.
Thanks you so much

**Comment by tranducanhbk** [comment]
Could u share to me the loss function and forward of RotationNet. I am still reach to good results.
Many thanks you

**Comment by mks0601** [comment]
You can see this repo: https://github.com/mks0601/Hand4Whole_RELEASE

**Comment by tranducanhbk** [comment]
I alread follow from this repo but there are 2 things i don't clear:
https://github.com/mks0601/Hand4Whole_RELEASE/blob/2f7a608cb05cf586d1e80c76e507d0802a6c13f0/main/config.py#L50
1) you calculated 2D joint loss in 256x256 space. so  self.output_hm_shape = (8, 8, 6) need to change to (256,256,256) right?
2) because i use 2D image so the input don't have Z axis so what value i need to set for Z axis of input before feed it to Rotation forwad funtion https://github.com/mks0601/Hand4Whole_RELEASE/blob/2f7a608cb05cf586d1e80c76e507d0802a6c13f0/common/nets/module.py#L58

**Comment by mks0601** [comment]
1. For the network of NeuralAnnot, I calculated the 2D loss in 256x256 space. For the network of Hand4Whole, I calcualted the 2D loss in 8x8. The networks of two works are different.
2. joint_coord_img is estimated from PositionNet.

**Comment by tranducanhbk** [comment]
I strictly follow what you guide but can not get good result.
Below are losses i caculate for backward
loss['joint_img'] = self.coord_loss(joint_img, smpl.reduce_joint_set(targets['joint_img'])/32., smpl.reduce_joint_set(meta_info['joint_trunc']), meta_info['is_3D']) for 2d loss of positionNet (in 8x8 space)
 loss['joint_proj'] = self.coord_loss(joint_proj, targets['joint_img'][:,:,:2], meta_info['joint_trunc']) for 2D loss of RatationNet (in 256x256 sapce 
loss['regularizer_body'] = self.regular_loss(smpl_body_pose)
loss['regularizer_shape'] = self.regular_loss(smpl_shape)

Class RegularizeLoss(nn.Module):
    def __init__(self):
        super(RegularizeLoss, self).__init__()

    def forward(self, smpl_para):
        loss = torch.pow(smpl_para,2)
        return loss

class CoordLoss(nn.Module):
    def __init__(self):
        super(CoordLoss, self).__init__()

    def forward(self, coord_out, coord_gt, valid, is_3D=None):
        loss = torch.abs(coord_out - coord_gt) * valid
        if is_3D is not None:
            loss_z = loss[:,:,2:] * is_3D[:,None,None].float()
            loss = torch.cat((loss[:,:,:2], loss_z),2)
        return loss
Please tell me what issue i got.
this is results
![render_original_img_body](https://user-images.githubusercontent.com/54969565/171660476-b309f6c2-ff83-490c-9281-214e512f02a1.jpg)

**Comment by mks0601** [comment]
Are you using VPoser? As written in the paper, VPoser is helpful.

**Comment by tranducanhbk** [comment]
I did not use VPoser. Maybe it is big misktake i got. Thank you for your advise

**Comment by tranducanhbk** [comment]
I used Vposer to decode SMPL pose from 32 dimenstions but still not got good results.
Below is code to get  root_pose, body_pose, shape_param, cam_param
# predict 32 dismention for vposer
self.vposer_out = make_linear_layers([2048, 32], relu_final=False)
# predict 63 body pose of smpl from vposer  32 vector
self.vposer, _  = load_model(cfg.vposer_expr_dir, model_code=VPoser,remove_words_in_model_weights='vp_model.', disable_grad=True)
# predict global rotaions of smpl 
self.root_pose_out = make_linear_layers([2048, 3], relu_final=False)
# predict shape parameters of smpl
 self.shape_out = make_linear_layers([2048,smpl.shape_param_dim], relu_final=False)
Forward fuction.
#shape_param
shape_param = self.shape_out(img_feat.mean((2,3)))
 # camera parameter
cam_param = self.cam_out(img_feat.mean((2,3)))
# vposer para 
 vposer_para = self.vposer_out(img_feat.mean((2,3)))
# body pose
 body_pose = self.vposer.decode(vposer_para)['pose_body'].contiguous().view(-1, 63)
 # root pose
 root_pose = self.root_pose_out(img_feat.mean((2,3)))
return root_pose, body_pose, shape_param, cam_param

Loss function
loss['joint_proj'] = self.coord_loss(joint_proj, targets['joint_img'][:,:,:2], meta_info['joint_trunc'])
loss['regularizer_body'] = self.regular_loss(smpl_body_pose)
loss['regularizer_shape'] = self.regular_loss(smpl_shape)

I already caculate 2D loss in 256x256 space and set L2  reg =1 but loss converges around loss_joint_proj: 7.6711 loss_regularizer_body: 1.4312 loss_regularizer_shape: 0.0089

![render_original_img_body](https://user-images.githubusercontent.com/54969565/172032790-eb74c4d6-2a61-4aae-b942-d23cb1f6996a.jpg)
I wonder what i got mistake now

**Comment by tranducanhbk** [comment]
I check in Vposer doc: "On the other hand, body poZ, VPoser's latent space representation for SMPL body, has in total 32 elements with a spherical Gaussian distribution. This means if one samples a 32 dimensional random vector from a Normal distribution and pass it through VPoser's decoder the result would be a viable human joint configuration in axis-angle representation"
I told Output of Linear layer should follow Gaussian distribution, can u share function for this

**Comment by mks0601** [comment]
You should apply the L2 regularizer to the vposer code. All my previous and current answers are included in the paper. Please read the paper carefully for your implementation.

**Comment by tranducanhbk** [comment]
Finally i got better result 
![correct](https://user-images.githubusercontent.com/54969565/172539914-3d9a3f45-550e-4661-9468-14ef2a3262c3.jpg)
but some case it still wrong
![wrong](https://user-images.githubusercontent.com/54969565/172539951-965d82f7-0a4f-43a7-9640-94396fd53a83.jpg)

**Comment by mks0601** [comment]
You'd better use SMPL instead of SMPLX as you do not have hand/face annotations. Also, try stronger L2 regularizer weight.

**Comment by tranducanhbk** [comment]
I am using SMPL.
Stronger L2 regularizer weight mean increase L2 regularizer weight right?

**Comment by mks0601** [comment]
Yes. As you can see qualitative results of the papers or provided 3D pseudo-GTs of MSCOCO in this repo, the 3D pseudo-GTs are quite good. There should be some problem in your implementation :(

**Comment by tranducanhbk** [comment]
Yes, I think so. I strickly follow your paper and now i don't know what problem I got.
If u can share some log of training and detail model architect it will helpful for me

**Comment by tranducanhbk** [comment]
this is source code base on your HAND4WHOLE ( now only for body)
https://github.com/tranducanhbk/NeuralAnnot
if u find any issue please tell me

**Comment by tranducanhbk** [comment]
Could u show some training log?

**Comment by tranducanhbk** [comment]
Finally  i can got good result.
Your paper now mention about feed one images for one time generate 3D GT. 
![0_render_original_img_body](https://user-images.githubusercontent.com/54969565/173055834-da99910c-3c68-4f82-9410-aa2e74bc239e.jpg)

**Comment by tranducanhbk** [comment]
Hi mks1, 
Now, I can make 3D Gt from 2D images, but some it generate wrong 3D GT with same file.
all configuration parameter is same but sometime 3D GT is correct and sometimes is not correct
This is correct 
![left_0178](https://user-images.githubusercontent.com/54969565/174039207-31038977-6bcc-4bc2-b92d-34ad9e967751.jpg)
 And this is  incorrect
![left_0178](https://user-images.githubusercontent.com/54969565/174038858-f4fab1cb-2e5c-42c7-a31e-9195fc6fdc0e.jpg)

Have u face this issuse before?

**Comment by nguyenviettuan96** [comment]
I had the same issue with @tranducanhbk. @mks0601 Do you have any idea to solve this problem?

**Comment by mks0601** [comment]
You mean the results are different when you train your network again?

**Comment by tranducanhbk** [comment]
Yes
sometimes it generate a different result

**Comment by mks0601** [comment]
Maybe because of random initialization and shuffling of the datasets

**Comment by tranducanhbk** [comment]
I create datasets with only one image ( images I need to make 3D GT)

**Comment by nguyenviettuan96** [comment]
> You mean the results are different when you train your network again?

yes

**Comment by EveningLin** [comment]
the result seems bad in MSCOCO  dataset. @mks0601 Do you have any idea to solve this problem?
![10](https://user-images.githubusercontent.com/110521494/202348601-5ec77bce-27e8-428f-91e8-218e911e2ea2.png)
![9](https://user-images.githubusercontent.com/110521494/202348606-aeddb1b5-b73d-4d3f-927b-736d2ef46232.png)

**Comment by mks0601** [comment]
Some samples can have had fit. Those are especially can suffer from the depth ambiguity.

**Comment by EveningLin** [comment]
@mks0601 You mean that not every piece of 2D dataset can achieve the appropriate effect.  I tested about 20 mscoco datasets in the morning, and the results were not very good.  i used the code provided by @tranducanhbk 
 could you please realse your training /test code@mks0601 ?
